File: python/mandatekit/mandate.py
"""
The mandate data model.

A mandate is a plain JSON-serializable dict so it travels over any transport and
signs deterministically. This module builds well-formed mandates, exposes the
JSON Schema, and provides a dependency-free structural validator.

Shape (tracks the AP2 Verifiable Intent draft; field names are MandateKit's own
until the spec finalizes):

    {
      "version": "mandatekit/v0",
      "spec": "AP2-draft-2026Q2",
      "mandate_id": "uuid4",
      "subject": {"agent_id": "..."},
      "issued_at": "2026-06-05T12:00:00Z",
      "expires_at": "2026-06-30T23:59:59Z",
      "constraints": {
        "intent": "buy running shoes",          # natural-language intent basket
        "categories": ["apparel"],               # allowed merchant categories
        "merchants": {"allow": [...], "deny": [...]},   # optional
        "max_amount": {"value": 500, "currency": "USD"}   # value: integer, caller-chosen unit
      }
    }
"""

import uuid
from datetime import datetime, timedelta, timezone
from typing import Dict, List, Optional

VERSION = "mandatekit/v0"
SPEC = "AP2-draft-2026Q2"


def _now() -> datetime:
    return datetime.now(timezone.utc)


def iso(dt: datetime) -> str:
    """UTC ISO-8601 with a trailing Z, second precision."""
    return dt.astimezone(timezone.utc).replace(microsecond=0).isoformat().replace(
        "+00:00", "Z"
    )


def parse_iso(s: str) -> datetime:
    dt = datetime.fromisoformat(s.replace("Z", "+00:00"))
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    return dt


def build_mandate(
    *,
    agent_id: str,
    intent: Optional[str] = None,
    categories: Optional[List[str]] = None,
    merchants: Optional[Dict[str, List[str]]] = None,
    max_amount: Optional[Dict] = None,
    issued_at: Optional[datetime] = None,
    expires_at: Optional[datetime] = None,
    ttl_days: int = 30,
    mandate_id: Optional[str] = None,
) -> Dict:
    issued = issued_at or _now()
    expires = expires_at or (issued + timedelta(days=ttl_days))

    constraints: Dict = {}
    if intent:
        constraints["intent"] = intent
    if categories:
        constraints["categories"] = list(categories)
    if merchants and (merchants.get("allow") or merchants.get("deny")):
        constraints["merchants"] = {
            k: list(v) for k, v in merchants.items() if v
        }
    if max_amount:
        value = max_amount["value"]
        if not isinstance(value, int) or isinstance(value, bool):
            raise ValueError(
                "max_amount.value must be an integer (amounts are compared like-for-like; "
                "pick a unit, e.g. cents or whole units, and use it consistently); "
                "floats are rejected to keep signatures canonical across languages"
            )
        constraints["max_amount"] = {
            "value": value,
            "currency": max_amount.get("currency", "USD"),
        }

    return {
        "version": VERSION,
        "spec": SPEC,
        "mandate_id": mandate_id or str(uuid.uuid4()),
        "subject": {"agent_id": agent_id},
        "issued_at": iso(issued),
        "expires_at": iso(expires),
        "constraints": constraints,
    }


MANDATE_SCHEMA: Dict = {
    "$schema": "https://json-schema.org/draft/2020-12/schema",
    "title": "MandateKit mandate (v0)",
    "type": "object",
    "required": [
        "version",
        "spec",
        "mandate_id",
        "subject",
        "issued_at",
        "expires_at",
        "constraints",
    ],
    "properties": {
        "version": {"const": VERSION},
        "spec": {"type": "string"},
        "mandate_id": {"type": "string"},
        "subject": {
            "type": "object",
            "required": ["agent_id"],
            "properties": {"agent_id": {"type": "string"}},
        },
        "issued_at": {"type": "string", "format": "date-time"},
        "expires_at": {"type": "string", "format": "date-time"},
        "constraints": {
            "type": "object",
            "properties": {
                "intent": {"type": "string"},
                "categories": {"type": "array", "items": {"type": "string"}},
                "merchants": {
                    "type": "object",
                    "properties": {
                        "allow": {"type": "array", "items": {"type": "string"}},
                        "deny": {"type": "array", "items": {"type": "string"}},
                    },
                },
                "max_amount": {
                    "type": "object",
                    "required": ["value", "currency"],
                    "properties": {
                        "value": {"type": "integer", "description": "caller-chosen unit, used consistently for cap and transaction"},
                        "currency": {"type": "string"},
                    },
                },
            },
        },
    },
}


def validate(mandate: Dict) -> List[str]:
    """Dependency-free structural check. Returns a list of problems ([] == valid)."""
    errors: List[str] = []
    for field in MANDATE_SCHEMA["required"]:
        if field not in mandate:
            errors.append(f"missing required field: {field}")
    if mandate.get("version") != VERSION:
        errors.append(f"version must be {VERSION!r}")
    subject = mandate.get("subject")
    if not isinstance(subject, dict) or not subject.get("agent_id"):
        errors.append("subject.agent_id is required")
    for field in ("issued_at", "expires_at"):
        if field in mandate:
            try:
                parse_iso(mandate[field])
            except Exception:
                errors.append(f"{field} is not a valid ISO-8601 datetime")
    c = mandate.get("constraints")
    if not isinstance(c, dict):
        errors.append("constraints must be an object")
    else:
        ma = c.get("max_amount")
        if ma is not None and (
            not isinstance(ma, dict)
            or not isinstance(ma.get("value"), int)
            or isinstance(ma.get("value"), bool)
            or not ma.get("currency")
        ):
            errors.append("max_amount must be {value:integer, currency:string}")
    return errors

File: typescript/src/mandate.ts
/**
 * The mandate data model (TypeScript port). Mirrors the Python module's shape
 * exactly so mandates are interchangeable across the two SDKs.
 */

import { randomUUID } from "node:crypto";

export const VERSION = "mandatekit/v0";
export const SPEC = "AP2-draft-2026Q2";

export interface Amount {
  value: number;
  currency: string;
}

export interface Constraints {
  intent?: string;
  categories?: string[];
  merchants?: { allow?: string[]; deny?: string[] };
  max_amount?: Amount;
}

export interface Mandate {
  version: string;
  spec: string;
  mandate_id: string;
  subject: { agent_id: string };
  issued_at: string;
  expires_at: string;
  constraints: Constraints;
}

export function iso(d: Date): string {
  return d.toISOString().replace(/\.\d{3}Z$/, "Z");
}

export function parseIso(s: string): Date {
  return new Date(s);
}

export interface BuildOptions {
  agentId: string;
  intent?: string;
  categories?: string[];
  merchants?: { allow?: string[]; deny?: string[] };
  maxAmount?: Amount;
  issuedAt?: Date;
  expiresAt?: Date;
  ttlDays?: number;
  mandateId?: string;
}

export function buildMandate(opts: BuildOptions): Mandate {
  const issued = opts.issuedAt ?? new Date();
  const ttlDays = opts.ttlDays ?? 30;
  const expires =
    opts.expiresAt ?? new Date(issued.getTime() + ttlDays * 86400000);

  const constraints: Constraints = {};
  if (opts.intent) constraints.intent = opts.intent;
  if (opts.categories && opts.categories.length) {
    constraints.categories = [...opts.categories];
  }
  if (opts.merchants && (opts.merchants.allow?.length || opts.merchants.deny?.length)) {
    constraints.merchants = {};
    if (opts.merchants.allow?.length) constraints.merchants.allow = [...opts.merchants.allow];
    if (opts.merchants.deny?.length) constraints.merchants.deny = [...opts.merchants.deny];
  }
  if (opts.maxAmount) {
    if (!Number.isInteger(opts.maxAmount.value)) {
      throw new Error(
        "maxAmount.value must be an integer (amounts are compared like-for-like; " +
          "pick a unit, e.g. cents or whole units, and use it consistently); " +
          "floats are rejected to keep signatures canonical across languages",
      );
    }
    constraints.max_amount = {
      value: opts.maxAmount.value,
      currency: opts.maxAmount.currency ?? "USD",
    };
  }

  return {
    version: VERSION,
    spec: SPEC,
    mandate_id: opts.mandateId ?? randomUUID(),
    subject: { agent_id: opts.agentId },
    issued_at: iso(issued),
    expires_at: iso(expires),
    constraints,
  };
}

export function validate(m: Mandate): string[] {
  const errors: string[] = [];
  const required = [
    "version",
    "spec",
    "mandate_id",
    "subject",
    "issued_at",
    "expires_at",
    "constraints",
  ];
  for (const f of required) {
    if (!(f in (m as unknown as Record<string, unknown>))) errors.push(`missing required field: ${f}`);
  }
  if (m.version !== VERSION) errors.push(`version must be '${VERSION}'`);
  if (!m.subject || !m.subject.agent_id) errors.push("subject.agent_id is required");
  for (const f of ["issued_at", "expires_at"] as const) {
    if (m[f] && Number.isNaN(Date.parse(m[f]))) {
      errors.push(`${f} is not a valid ISO-8601 datetime`);
    }
  }
  const ma = m.constraints?.max_amount;
  if (ma && (!Number.isInteger(ma.value) || !ma.currency)) {
    errors.push("max_amount must be {value:integer, currency:string}");
  }
  return errors;
}
